src/main/java/org/antlr/intellij/plugin/validation/GrammarFileInspector.java

```diff
@@ -0,0 +1,77 @@
+package org.antlr.intellij.plugin.validation;
+
+import com.intellij.codeInspection.LocalQuickFix;
+import com.intellij.codeInspection.ProblemsHolder;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiRecursiveElementVisitor;
+import org.antlr.intellij.adaptor.lexer.TokenIElementType;
+import org.antlr.intellij.plugin.psi.LexerRuleRefNode;
+import org.antlr.runtime.CommonToken;
+import org.antlr.v4.tool.ErrorType;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+public class GrammarFileInspector extends PsiRecursiveElementVisitor {
+
+    private final ProblemsHolder holder;
+
+    private final List<GrammarIssue> grammarIssues;
+
+    public GrammarFileInspector(ProblemsHolder holder, List<GrammarIssue> grammarIssues) {
+        this.holder = Objects.requireNonNull(holder);
+        this.grammarIssues = Objects.requireNonNull(grammarIssues);
+    }
+
+    @Override
+    public void visitElement(PsiElement element) {
+        if (element instanceof LexerRuleRefNode) {
+            handleLeafElement((LexerRuleRefNode) element, holder);
+        }
+        super.visitElement(element);
+    }
+
+    private void handleLeafElement(LexerRuleRefNode element, ProblemsHolder holder) {
+        if (!(element.getElementType() instanceof TokenIElementType)) {
+            return;
+        }
+
+        TokenIElementType tokenIElementType = (TokenIElementType) element.getElementType();
+        if (!"TOKEN_REF".equalsIgnoreCase(org.antlr.intellij.plugin.parser.ANTLRv4Lexer.VOCABULARY.getDisplayName(tokenIElementType.getANTLRTokenType()))) {
+            return;
+        }
+
+        for (GrammarIssue grammarIssue : grammarIssues) {
+            boolean isAmongOffendingTokens = isAmongOffendingTokens(element, grammarIssue);
+            if (isAmongOffendingTokens) {
+                Optional<LocalQuickFix> localQuickFix = getFix(element, grammarIssue.getMsg().getErrorType());
+
+                if (localQuickFix.isPresent()) {
```

Comment:
The warning is registered twice (once by `ANTLRv4ExternalAnnotator`, and once by your code):

![image](https://user-images.githubusercontent.com/281528/60462234-0b70d380-9c49-11e9-911e-6f49c5f34457.png)

One possible solution would be to modify `ANTRLv4Inspection` to run only when `isOnTheFly` is `true`, and let `ANTLRv4ExternalAnnotator` register the quick fix. Another benefit is that `GrammarIssuesCollector.collectGrammarIssues()` will be called once instead of twice when the current editor is modified.

Reply:
I was thinking about replacing ANTLRv4ExternalAnnotator completely with ANTRLv4Inspection. As far as I'm concerned Annotators does not allow to bind quick fix to code. What do you think?

Reply:
Actually you can register quick fixes if you retrieve the result of `createXxxAnnotation()`:

```java
Annotation errorAnnotation = holder.createErrorAnnotation(range, issue.annotation);
errorAnnotation.registerFix(...);
```

Reply:
I couldn't remember the big differences between inspections and annotators, but found a nice explanation [on this page](https://www.jetbrains.org/intellij/sdk/docs/reference_guide/custom_language_support/code_inspections_and_intentions.html?search=inspec). 

Inspections can be disabled, which means that if we replace the existing annotator with a single inspection, then disabling it will disable the whole highlighting.

On the other hand, creating multiple inspections will probably provide poor performance because the problems are computed by the ANTLR tool. We can't possibly call the tool once per inspection.

I'm not sure running your newly created quick fix in batch mode is relevant, because the user will probably have to edit each automatic fix manually to replace things like `M Y T O K E N`.

To summarize: it seems to me that the current annotator is the correct way to highlight all these problems because they come from a single source (the output of the ANTLR tool). We should probably modify it to register quick fixes when they exist.

We can still add inspections later on for other kinds of problems for which an automatic fix does not require further editing.

WDYT?

Reply:
After reading your explanation i think that i will try to stick with annotator and remove inspection.

Reply:
I have removed inspections and registered fix with annotation.